### DISCUSSION TITLE
What is the best way to migrate data when upgrading from version 2.2.16 to 2.4.8?

### DISCUSSION THREAD

**Discussion by mmahao** [milvus-io/milvus#35544] [Q&A and General discussion]
The current production environment is using version 2.2.16, and we want to upgrade to version 2.4.8 to leverage new features and simplify development using MilvusPlus. What methods or tools can be used to migrate collections and data seamlessly and efficiently without any data loss?

**Comment by yhmo** [comment]
Normally, v2.2.16 can be upgraded to v2.3 or v2.4 without special operations.

In normal cases, you can simply update your Milvus to the latest by changing its image. However, note that there will be a certain downtime when upgrading Milvus in this way. For a production environment, you might want a rolling upgrade. 

Read the doc for more details:
Upgrade to v2.3.x: https://milvus.io/docs/v2.3.x/upgrade_milvus_cluster-operator.md
Upgrade to v2.4.x: https://milvus.io/docs/upgrade_milvus_cluster-operator.md

**Reply by mmahao** [reply]
thanks

**Reply by mmahao** [reply]
By the way, since Milvus supports rolling upgrades, there will still be some downtime when upgrading Java projects that reference the Milvus SDK, right? This is because there are compatibility issues between the SDK and the Milvus version during the upgrade, correct?

**Reply by yhmo** [reply]
If you want to use the new features such as sparse/float16 vector, Array data type, hybrid search, etc, then you have to upgrade your Java project with the latest Java SDk.

**Comment by xiaofan-luan** [comment]
I would recommend to upgrade 2.2.16 to 2.3.21 and then do 2.4.
Use 2.2 Java SDK to 2.4 is actually compatible

**Reply by mmahao** [reply]
If possible, I would like to understand why it's recommended to upgrade to 2.3.21 first?

**Reply by xiaofan-luan** [reply]
we don't test to upgrade data across two major releases. So there might be some risk here.

**Reply by mmahao** [reply]
ok
